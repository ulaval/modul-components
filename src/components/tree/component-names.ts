export const TREE_NODE_NAME: string = 'm-tree-node';
export const TREE_ICON_NAME: string = 'm-tree-icon';
