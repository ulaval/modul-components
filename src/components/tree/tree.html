<div class="m-tree">
    <m-message v-if="errorTree" class="m-tree__error" state="warning" skin="light" ref="error-tree-txt">
        {{'m-tree:error' | f-m-i18n}}
    </m-message>
    <p v-if="propTreeEmpty" class="m-tree__empty" ref="empty-tree-txt">
        {{'m-tree:empty' | f-m-i18n}}
    </p>
    <ul v-else class="m-tree__content">
        <m-tree-node ref="tree-node" v-for="node in tree" :key="node.id" :selectable="selectable" :open.sync="node.open" :node="node" :selected-nodes="propSelectedNodes"
            :icons="icons" @click="onClick"></m-tree-node>
    </ul>
</div>
