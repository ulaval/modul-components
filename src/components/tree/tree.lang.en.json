{
    "m-tree": {
        "empty": "No content",
        "error": "Some items could not be displayed because of a technical problem."
    }
}
