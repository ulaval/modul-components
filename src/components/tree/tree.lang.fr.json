{
    "m-tree": {
        "empty": "Aucun contenu",
        "error": "Certains éléments n'ont pas pu être affichés dû à un problème technique."
    }
}
