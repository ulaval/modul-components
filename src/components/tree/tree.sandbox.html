<div>
    <h2>Default tree</h2>
    <div style="width: 50%; float:left;">
        <h2>No selection</h2>
        <p>
            <m-tree :tree="tree" selectionMode="0" ></m-tree>
        </p>
    </div>

    <div style="width: 50%; float:left;">
        <h2>Single node selection</h2>
        <p>
            <m-tree :tree="tree" :selectedNodes="currentFile2" @select="onSelect"></m-tree>
        </p>
    </div>

    <div style="clear: both;"></div>

    <h2>File tree</h2>
    <div style="width: 50%; float:left;">
        <h3>No file selection</h3>
        <p>
            <m-tree :tree="fileTree" selectionMode="0" :icons="true"></m-tree>
        </p>
    </div>

    <div style="width: 50%; float:left;">
        <h3>Single file selection</h3>
        <p>
            <m-tree :tree="fileTree" selectionMode="1" :selectedNodes="currentFile" :icons="true" @select="onSelect"></m-tree>
        </p>
    </div>

    <div style="clear: both;"></div>

    <h2>Other cases</h2>

    <div style="width: 50%; float:left;">
        <h3>Nodes error</h3>
        <p>
            <m-tree :tree="errorTree"></m-tree>
        </p>
    </div>

    <div style="width: 50%; float:left;">
        <h3>Empty tree</h3>
        <p>
            <m-tree :tree="emptyTree"></m-tree>
        </p>
    </div>

    <div style="clear: both;"></div>

    <div style="width: 50%; float:left;">
        <h3>Selected node not found</h3>
        <span>Console error</span>
        <p>
            <m-tree :tree="tree" :selectedNodes="wrongCurrentFile" @select="onSelect"></m-tree>
        </p>
    </div>

</div>
