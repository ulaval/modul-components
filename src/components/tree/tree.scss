@import '../../styles/abstracts/variables';

.m-tree {
    overflow: auto;

    &__error {
        .m-icon {
            color: $m-color--error-dark;
        }

        + .m-tree__content {
            margin-top: $m-margin;
        }
    }

    &__content {
        margin-top: 0;
        margin-bottom: 0;
    }
}
