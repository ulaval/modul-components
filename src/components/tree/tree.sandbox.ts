import Vue, { PluginObject } from 'vue';
import Component from 'vue-class-component';

import { TREE_NAME } from '../component-names';
import { TreeNode } from './tree';
import WithRender from './tree.sandbox.html';

export class Tree {
    id: string;
    label?: string;
    uselessData?: string;
    hasChildren?: boolean;
}

@WithRender
@Component
export class MRootTreeSandbox extends Vue {
    public currentFile: string[] = ['/folder 1/folder 2/index.html'];
    public currentFile2: string[] = ['/1/2'];
    public wrongCurrentFile: string[] = ['/3/4'];

    public emptyTree: TreeNode[] = [];

    public tree: TreeNode[] = [
        {
            id: '1',
            label: 'Title 1',
            hasChildren: true,
            children: [
                {
                    id: '2',
                    label: 'Subtitle 1'
                },
                {
                    id: '3',
                    label: 'Subtitle 2',
                    children: [
                        {
                            id: '2',
                            label: 'Subtitle 1'
                        },
                        {
                            id: '3',
                            label: 'Subtitle 2',
                            hasChildren: true
                        },
                        {
                            id: '4',
                            label: 'Subtitle 3'
                        },
                        {
                            id: '5',
                            label: 'Subtitle 4'
                        }
                    ]
                },
                {
                    id: '4',
                    label: 'Subtitle 3'
                },
                {
                    id: '5',
                    label: 'Subtitle 4'
                }
            ]
        },
        {
            id: '6',
            label: 'Title 2'
        },
        {
            id: '7',
            label: 'Title 3',
            children: [
                {
                    id: '1',
                    label: 'Subtitle 1'
                },
                {
                    id: '2',
                    label: 'Subtitle 2'
                }
            ]
        }
    ];

    public errorTree: TreeNode[] = [
        {
            id: '1',
            label: 'Title 1',
            hasChildren: true,
            children: [
                {
                    id: '',
                    label: 'Subtitle 1'
                },
                {
                    id: '3',
                    label: 'Subtitle 2'
                },
                {
                    id: '4',
                    label: 'Subtitle 3'
                },
                {
                    id: '5',
                    label: 'Subtitle 4'
                }
            ]
        }
    ];

    public fileTree: TreeNode[] = [
        {
            id: 'folder 1',
            label: 'folder 1',
            hasChildren: true,
            children: [
                {
                    id: 'index.html',
                    label: 'index.html'
                },
                {
                    id: 'menu.html'
                },
                {
                    id: 'folder 2',
                    label: 'folder 2',
                    hasChildren: true,
                    children: [
                        {
                            id: 'index.html',
                            label: 'index.html'
                        },
                        {
                            id: 'folder 3',
                            label: 'folder 3',
                            hasChildren: true,
                            children: []
                        }
                    ]
                }
            ]
        },
        {
            id: 'poney.jpg',
            label: 'poney.jpg'
        },
        {
            id: 'folder 4',
            label: 'folder 4',
            hasChildren: true
        },
        {
            id: 'big folder',
            label: 'big folder',
            hasChildren: true,
            children: [
                {
                    id: 'folder 1',
                    label: 'folder 1',
                    hasChildren: true,
                    children: [
                        {
                            id: 'index.html',
                            label: 'index.html'
                        },
                        {
                            id: 'menu.html'
                        },
                        {
                            id: 'folder 2',
                            label: 'folder 2',
                            hasChildren: true,
                            children: [
                                {
                                    id: 'index.html',
                                    label: 'index.html'
                                },
                                {
                                    id: 'folder 3',
                                    label: 'folder 3',
                                    hasChildren: true,
                                    children: [
                                        {
                                            id: 'folder 1',
                                            label: 'folder 1',
                                            hasChildren: true,
                                            children: [
                                                {
                                                    id: 'index.html',
                                                    label: 'index.html'
                                                },
                                                {
                                                    id: 'menu.html'
                                                },
                                                {
                                                    id: 'folder 2',
                                                    label: 'folder 2',
                                                    hasChildren: true,
                                                    children: [
                                                        {
                                                            id: 'index.html',
                                                            label: 'index.html'
                                                        },
                                                        {
                                                            id: 'folder 3',
                                                            label: 'folder 3',
                                                            hasChildren: true,
                                                            children: [
                                                                {
                                                                    id: 'folder 1',
                                                                    label: 'folder 1',
                                                                    hasChildren: true,
                                                                    children: [
                                                                        {
                                                                            id: 'index.html',
                                                                            label: 'index.html'
                                                                        },
                                                                        {
                                                                            id: 'menu.html'
                                                                        },
                                                                        {
                                                                            id: 'folder 2',
                                                                            label: 'folder 2',
                                                                            hasChildren: true,
                                                                            children: [
                                                                                {
                                                                                    id: 'index.html',
                                                                                    label: 'index.html'
                                                                                },
                                                                                {
                                                                                    id: 'folder 3',
                                                                                    label: 'folder 3',
                                                                                    hasChildren: true,
                                                                                    children: []
                                                                                }
                                                                            ]
                                                                        }
                                                                    ]
                                                                }
                                                            ]
                                                        }
                                                    ]
                                                }
                                            ]
                                        }
                                    ]
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ];

    public onSelect(): void {
        console.error('modUL - New file selected');
    }

}

const RootTreeSandboxPlugin: PluginObject<any> = {
    install(v, options): void {
        v.component(`${TREE_NAME}-sandbox`, MRootTreeSandbox);
    }
};

export default RootTreeSandboxPlugin;
