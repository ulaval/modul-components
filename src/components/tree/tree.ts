import { PluginObject } from 'vue';
import Component from 'vue-class-component';
import { Prop, Watch } from 'vue-property-decorator';

import { ModulVue } from '../../utils/vue/vue';
import { TREE_NAME } from '../component-names';
import I18nPlugin from '../i18n/i18n';
import TreeNodePlugin from '../tree-node/tree-node';
import WithRender from './tree.html?style=./tree.scss';

export interface TreeNode {
    id: string;
    label?: string;
    open?: boolean;
    children?: TreeNode[];
    hasChildren?: boolean;
    data?: any;
}

export enum MSelectionMode {
    None = '0',
    Single = '1',
    Multiple = '2'
}

@WithRender
@Component
export class MTree extends ModulVue {
    @Prop()
    public tree: TreeNode[];

    @Prop({
        default: MSelectionMode.Single,
        validator: value =>
            value === MSelectionMode.None ||
            value === MSelectionMode.Single ||
            value === MSelectionMode.Multiple
    })
    public selectionMode: MSelectionMode;

    @Prop()
    public selectedNodes: string[];

    @Prop()
    public icons: boolean;

    public propSelectedNodes: string[] = this.selectedNodes || [];
    public errorTree: boolean = false;

    private selectedNodesFound: string[] = [];

    public onClick(path: string): void {
        if (this.propSelectedNodes.indexOf(path) === -1) {
            if (this.selectionMode === MSelectionMode.Multiple) {
                this.propSelectedNodes.push(path);
            } else {
                this.propSelectedNodes = [path];
            }
        } else if (this.selectionMode === MSelectionMode.Multiple) {
            this.propSelectedNodes.splice(this.propSelectedNodes.indexOf(path), 1);
        }
        this.$emit('select', path);
    }

    public get propTreeEmpty(): boolean {
        return !this.tree.length;
    }

    public get selectable(): boolean {
        return this.selectionMode !== MSelectionMode.None;
    }

    protected created(): void {
        this.browseTree();
    }

    @Watch('tree')
    private browseTree(): void {
        this.errorTree = false;
        this.tree.forEach(node => {
            this.browseNode(node);
        });
        this.propSelectedNodes.forEach(selectedNode => {
            if (this.selectedNodesFound.indexOf(selectedNode) === -1) {
                console.error(`modUL - The selected node was not found: "${selectedNode}"`);
            }
        });
    }

    private browseNode(node: TreeNode, path: string = ''): void {
        if (node.id.trim() === '') {
            this.errorTree = true;
        }
        let currentPath: string = path + '/' + node.id;
        if (this.propSelectedNodes.indexOf(currentPath) !== -1) {
            this.selectedNodesFound.push(currentPath);
        }
        if (node.children) {
            node.children.forEach(childNode => {
                this.browseNode(childNode, currentPath);
            });
        }
    }
}

const TreePlugin: PluginObject<any> = {
    install(v, options): void {
        v.prototype.$log.debug(TREE_NAME, 'plugin.install');
        v.use(TreeNodePlugin);
        v.use(I18nPlugin);
        v.component(TREE_NAME, MTree);
    }
};

export default TreePlugin;
