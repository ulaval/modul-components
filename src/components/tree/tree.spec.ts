import { RefSelector, shallow, Wrapper } from '@vue/test-utils';

import { renderComponent } from '../../../tests/helpers/render';
import { MSelectionMode, MTree, TreeNode } from './tree';

const TREE_NODE_REF: RefSelector = { ref: 'tree-node' };
const EMPTY_TREE_REF: RefSelector = { ref: 'empty-tree-txt' };
const ERROR_TREE_REF: RefSelector = { ref: 'error-tree-txt' };

const SELECTED_NODES: string[] = ['/medias/Videos'];
const SELECTED_NODES_INVALID: string[] = ['/medias/Videos/video-dog.mov'];
const NEW_TREE_NODE_SELECTED: string[] = ['/index.html'];

const EMPTY_TREE: TreeNode[] = [];
const TREE_WITH_DATA: TreeNode[] = [
    {
        label: 'index.html',
        id: 'index.html'
    },
    {
        label: 'Medias',
        id: 'medias',
        children: [
            {
                label: 'Videos',
                id: 'Videos'
            }
        ]
    }
];
const TREE_WITH_NIVALID_DATA: TreeNode[] = [
    {
        label: 'index.html',
        id: ''
    },
    {
        label: 'Medias',
        id: 'medias',
        children: [
            {
                label: 'Videos',
                id: 'Videos'
            }
        ]
    }
];

let wrapper: Wrapper<MTree>;

let tree: TreeNode[] = TREE_WITH_DATA;
let selectionMode: MSelectionMode = MSelectionMode.Single;
let selectedNodes: string[] = SELECTED_NODES;

afterEach(() => {
    tree = [];
    selectionMode = MSelectionMode.Single;
    selectedNodes = SELECTED_NODES;
});

const initializeShallowWrapper: any = () => {
    wrapper = shallow(MTree, {
        propsData: {
            tree,
            selectedNodes,
            selectionMode
        }
    });
};

describe(`MTree`, () => {

    describe(`Given an empty tree`, () => {

        beforeEach(() => {
            tree = EMPTY_TREE;
            initializeShallowWrapper();
        });

        it(`Should render correctly`, () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

        it(`Should be empty`, () => {
            expect(wrapper.vm.propTreeEmpty).toBeTruthy();
        });
    });

    describe(`Given a tree with some data`, () => {

        beforeEach(() => {
            tree = TREE_WITH_DATA;
            initializeShallowWrapper();
        });

        it(`Should render correctly`, () => {
            expect(renderComponent(wrapper.vm)).resolves.toMatchSnapshot();
        });

        it(`Should not be empty`, () => {
            expect(wrapper.vm.propTreeEmpty).toBeFalsy();
        });

        describe(`When a node is selected`, () => {

            beforeEach(() => {
                wrapper.vm.onClick(NEW_TREE_NODE_SELECTED[0]);
            });

            it(`Call onClick`, () => {
                wrapper.setMethods({ onClick: jest.fn() });
                wrapper.find(TREE_NODE_REF).trigger('click');

                expect(wrapper.vm.onClick).toHaveBeenCalled();
            });

            it(`Emit select`, () => {
                expect(wrapper.emitted('select')).toBeTruthy();
            });

            it(`A new node is selected`, () => {
                expect(wrapper.vm.propSelectedNodes).toEqual(NEW_TREE_NODE_SELECTED);
            });
        });

        describe(`When there is an error in the tree`, () => {

            beforeEach(() => {
                tree = TREE_WITH_NIVALID_DATA;
                initializeShallowWrapper();
            });

            it(`Should generate an error`, () => {
                expect(wrapper.vm.errorTree).toBeTruthy();
            });

            it(`Should show an error message`, () => {
                expect(wrapper.find(ERROR_TREE_REF).exists()).toBeTruthy();
            });

        });
    });
});
