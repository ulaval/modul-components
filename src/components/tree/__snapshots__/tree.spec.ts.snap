// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MTree Given a tree with some data Should render correctly 1`] = `
<div class="m-tree">
    <ul class="m-tree__content">
        <m-tree-node selectable="true" node="[object Object]" selected-nodes="/medias/Videos"></m-tree-node>
        <m-tree-node selectable="true" node="[object Object]" selected-nodes="/medias/Videos"></m-tree-node>
    </ul>
</div>
`;

exports[`MTree Given an empty tree Should render correctly 1`] = `
<div class="m-tree">
    <p class="m-tree__empty">
        m-tree:empty
    </p>
</div>
`;
